include/RED4ext/Callback.hpp

```diff
@@ -0,0 +1,382 @@
+
+    template<typename T>
+    using MoveFunc = CopyFunc<T>;
+
+    template<typename T>
+    using DestructFunc = void (*)(T* aTarget);
+
+    CallbackHandler(InvokeFunc<void> aInvoke, CopyFunc<void> aCopy, MoveFunc<void> aMove,
+                    DestructFunc<void> aDestruct) noexcept
+        : invoke(aInvoke)
+        , copy(aCopy)
+        , move(aMove)
+        , destruct(aDestruct)
+    {
+    }
+
+    template<typename T>
+    CallbackHandler(InvokeFunc<T> aInvoke, CopyFunc<T> aCopy, MoveFunc<T> aMove, DestructFunc<T> aDestruct) noexcept
+        : invoke(reinterpret_cast<InvokeFunc<void>>(aInvoke))
+        , copy(reinterpret_cast<CopyFunc<void>>(aCopy))
+        , move(reinterpret_cast<MoveFunc<void>>(aMove))
+        , destruct(reinterpret_cast<DestructFunc<void>>(aDestruct))
+    {
+    }
+
+    InvokeFunc<void> invoke;     // 00
+    CopyFunc<void> copy;         // 08
+    MoveFunc<void> move;         // 10
+    DestructFunc<void> destruct; // 18
+};
+RED4EXT_ASSERT_SIZE(CallbackHandler<void>, 0x20);
+RED4EXT_ASSERT_OFFSET(CallbackHandler<void>, invoke, 0x00);
+RED4EXT_ASSERT_OFFSET(CallbackHandler<void>, copy, 0x08);
+RED4EXT_ASSERT_OFFSET(CallbackHandler<void>, move, 0x10);
+RED4EXT_ASSERT_OFFSET(CallbackHandler<void>, destruct, 0x18);
+
+template<typename T, size_t InlineSize = DefaultFixedCallbackBufferSize>
+class Callback;
+
+template<typename R, typename... Args, size_t InlineSize>
+class Callback<R (*)(Args...), InlineSize>
+{
+public:
+    using HandlerPtr = CallbackHandler<R, Args...>*;
+
+    static_assert(InlineSize > sizeof(void*), "Buffer size can't be less than pointer size");
+
+    Callback(R (*aFunc)(Args...)) noexcept
+    {
+        using TargetType = Detail::UnboundFunctionTarget<R, Args...>;
+
+        static_assert(sizeof(TargetType) <= InlineSize, "Function size is too big for this callback");
+
+        TargetType target{aFunc};
+        InitializeHandler(&target);
+    }
+
+    template<typename C>
+    Callback(C* aContext, R (C::*aFunc)(Args...)) noexcept
+    {
+        using TargetType = Detail::MemberFunctionTarget<C, R, Args...>;
+
+        static_assert(sizeof(TargetType) <= InlineSize, "Function size is too big for this callback");
+
+        TargetType target{aContext, aFunc};
+        InitializeHandler(&target);
+    }
+
+    template<typename L>
+    requires Detail::IsClosure<L, R, Args...>
+    Callback(L&& aClosure) noexcept
+    {
+        using TargetType = Detail::ClosureTarget<L, R, Args...>;
+
+        static_assert(sizeof(TargetType) <= InlineSize, "Closure size is too big for this callback");
+
+        InitializeHandler(reinterpret_cast<TargetType*>(&aClosure));
+    }
+
+    Callback(const Callback& aOther) noexcept
+    {
+        CopyHandlerFrom(aOther.handler);
+        CopyTargetFrom(aOther.buffer);
+    }
+
+    Callback(Callback&& aOther) noexcept
+    {
+        MoveHandlerFrom(aOther.handler);
+        MoveTargetFrom(aOther.buffer);
+    }
+
```

Comment:
And in this case we have to destroy the current target too:
```cpp
    Callback& operator=(const Callback& aRhs)
    {
        if (this != std::addressof(aOther))
        {
            DestroyTarget();
            CopyHandlerFrom(aOther.handler);
            CopyTargetFrom(aOther.buffer);
        }

        return *this;
    }
```

Reply:
Why should we destroy it in case of copy?

Reply:
Because we are overwriting lhs buffer with the rhs one.
On construction, there is nothing in `this->buffer`.
But on assignment, there will be something in `this->buffer` since we are working with an initialized instance.

Reply:
Oh, true. For some reason I read that `DestroyTarget` is called on `rhs`.